### PR TITLE
Node: refine `Dict` type in favour of new compiler option `noUncheckedIndexedAccess`

### PR DESCRIPTION
**Pull request by OliverJAsh**
Supersedes https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51160

In https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43931, all index signatures were updated to include `undefined`, via a new helper called `Dict`. This was a workaround for https://github.com/microsoft/TypeScript/issues/13778. However, now that https://github.com/microsoft/TypeScript/issues/13778 is resolved and we have the new compiler option `noUncheckedIndexedAccess` at our disposal, this workaround is no longer necessary.

This PR removes `undefined` from `Dict` for TS versions 4.1 and greater (`noUncheckedIndexedAccess` was added in 4.1).

By making this change we also fix the following issue:

```ts
declare const dict: ParsedUrlQuery;

const value = dict['foo']; // string | string[] | undefined ✅ (when using `noUncheckedIndexedAccess`)

Object.values(dict).map((value) => {
    // Expected: string | string[];
    // Actual: string | string[] | undefined; ❌
    value;
});
```

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/node/index.d.ts
Comment (OliverJAsh) [inline types/node/index.d.ts:56]: My plan is to make this type specific to 4.2+, but to begin with I just wanted to test the versioning mechanism works.

When I try to run `npm run test-all` I get this error:

```
> definitely-typed@0.0.3 test-all /Users/oliverash/Development/DefinitelyTyped
> node node_modules/@definitelytyped/dtslint-runner/dist/index.js --path .

Using local DefinitelyTyped at .
Parsing definitions...
Found 7500 packages.
Parsing in 16 child process(es)...
Error: Unused file ./types/node/ts3.4/assert.d.ts (used files: ["index.d.ts","node-tests.ts","../test/assert.ts","../test/async_hooks.ts","../test/child_process.ts","../test/cluster.ts","../test/constants.ts","../test/crypto.ts","../test/dgram.ts","../test/dns.ts","../test/events.ts","../test/fs.ts","../test/global.ts","../test/http.ts","../test/http2.ts","../test/module.ts","../test/net.ts","../test/os.ts","../test/path.ts","../test/perf_hooks.ts","../test/process.ts","../test/readline.ts","../test/repl.ts","../test/stream.ts","../test/string_decoder.ts","../test/tls.ts","../test/tty.ts","../test/util.ts","../test/v8.ts","../test/vm.ts","../test/worker_threads.ts","../test/zlib.ts","tsconfig.json","tslint.json"])
```

Does anyone know why this is happening?

Comment (OliverJAsh) [reply types/node/index.d.ts:56]: Nevermind, I fixed it by moving the `declare` after the `reference`s. 🤷‍♂️

File: types/node/index.d.ts
Comment (ExE-Boss) [inline types/node/index.d.ts:63]: This should probably be moved into a `types/node/globals.nodejs.dict‑type.d.ts` file, just like how `types/node/globals.global.d.ts` is handled: <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/20efb58b919f1ff555b7137c57a337d20f90668f/types/node/globals.global.d.ts#L1>

File: types/node/ts3.4/index.d.ts
Comment (ExE-Boss) [inline types/node/ts3.4/index.d.ts:18]: ditto
```suggestion
/// <reference path="globals.nodejs.dict-type.d.ts" />
```

File: types/node/ts3.6/index.d.ts
Comment (ExE-Boss) [inline types/node/ts3.6/index.d.ts:17]: ditto
```suggestion
/// <reference path="../ts3.4/globals.nodejs.dict-type.d.ts" />
```

**Review by ExE-Boss — CHANGES_REQUESTED** [review]

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## concurrently ([<kbd>unpkg</kbd>](https://unpkg.com/browse/concurrently@latest/))
was missing the following properties:
1. concurrently
1. Logger
1. InputHandler
1. KillOnSignal
1. KillOthers

<details>
<summary>as well as these 13 other properties...</summary>
<p>LogError, LogExit, LogOutput, RestartProcess, concurrently, Logger, InputHandler, KillOnSignal, KillOthers, LogError, LogExit, LogOutput, RestartProcess</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 451a78c12418a841a890305a2bf8f51cf44d8969
</p>

File: types/node/ts4.0/tslint.json
Comment (OliverJAsh) [inline types/node/ts4.0/tslint.json:1]: I copied this from `types/node/tslint.json`. I wasn't sure whether to duplicate or use `extends`?

File: types/node/ts4.0/tslint.json
Comment (OliverJAsh) [inline types/node/ts4.0/tslint.json:11]: Without this I had linting errors. I copied this line from `types/node/ts3.6/tslint.json`.

File: types/node/ts4.0/tsconfig.json
Comment (OliverJAsh) [inline types/node/ts4.0/tsconfig.json:1]: I copied this from `types/node/tsconfig.json`. I wasn't sure whether to duplicate or use `extends`?

File: types/node/ts4.0/tsconfig.json
Comment (OliverJAsh) [inline types/node/ts4.0/tsconfig.json:9]: After I copied this from `types/node/tsconfig.json`, this was `es2018` but I had errors due to usage of `bigint`. I copied this line from `types/node/ts3.6/tsconfig.json`.

File: types/node/ts4.0/node-tests.ts
Comment (OliverJAsh) [inline types/node/ts4.0/node-tests.ts:1]: Originally I tried `import '../node-tests' but then `tsc` failed due to "expected errors" (`$ExpectError`). It looks like the other version folders don't bother running tests which contain expected errors, so I just copied what they were doing.

File: types/node/index.d.ts
Comment (OliverJAsh) [inline types/node/index.d.ts:59]: I wasn't sure whether to update the versions in the comments below. I don't really understand them because they are referencing a non-existent folder. There is no `ts3.5`.

File: types/node/ts4.0/base.d.ts
Comment (OliverJAsh) [inline types/node/ts4.0/base.d.ts:7]: I'm a bit confused by these comments but I just copied what I saw in `types/node/ts3.6/base.d.ts` and updated the version from `3.5` to `3.7`.

File: types/node/index.d.ts
Comment (rbuckton) [reply types/node/index.d.ts:59]: @sandersn switched most package definitions from referring to the oldest TS version in `index.d.ts` (with newer versions in `ts*` directories) to the latest TS version in `index.d.ts` (with older versions in `ts*` directories). It looks like this comment is just stale and needs to be updated.

File: types/node/ts4.0/base.d.ts
Comment (rbuckton) [reply types/node/ts4.0/base.d.ts:7]: These comments are just stale and need to be updated to match the new file layout related to `"typesVersions"`.

File: types/node/ts4.0/tslint.json
Comment (rbuckton) [reply types/node/ts4.0/tslint.json:10]: Where there a large number of linting errors? Generally its better to disable lint warnings in specific places (though the node definitions have never fully caught up to all of our lint rules.

**Comment by G-Rath** [comment]
I disagree with this change, at least in part: `noUncheckedIndexedAccess` is not like the `strict*` options in that it has clear pros and cons that make it a "x or y" choice rather than being a Strictly Better option (that's one of the reasons why it doesn't start with `strict`, and why its not enabled in the [default `tsconfigs`](https://github.com/tsconfig/bases/pull/40#issuecomment-764475433)).

With the current layout of the types you can mark things as defined using interface merging, which encourages you to explicitly define your types for things, i.e:

```
const encodeBasicCredentials = (username: string, password: string): string => {
  const credentials = Buffer.from(`${username}:${password}`).toString('base64');

  return `Basic ${credentials}`;
};

const credentials = encodeBasicCredentials(
  process.env.BASIC_AUTH_USERNAME,
  process.env.BASIC_AUTH_PASSWORD
);
```

Right now if I was to do to something like `BASIC_AUTH_USER`, TypeScript will tell me that it could be undefined, making me aware of the typo. However, this change will mean that'll only happen if `noUncheckedIndexedAccess` is enabled, which'll require me to make changes elsewhere in my codebase to account for conditions that `noUncheckedIndexedAccess` doesn't support as being sufficient for ensuring an index exists.

------

I'm not saying I'm against this *completely* - there may be types that shouldn't be `| undefined`, but I think types for properties like `process.env`, `headers` on Http requests, and other such places should retain their `| undefined` to  encourage more explicit and better typings.

**Comment by OliverJAsh** [comment]
@G-Rath Thanks for sharing your thoughts!

Adding `undefined` to an index signature solves the problem of "indexing doesn't return `undefined`" but it creates more problems when the index signature is used in other ways. For example

```ts
Object.values(process.env).map((value) => {
  // Expected: string;
  // Actual: string | undefined; ❌
  value;
});
```

or

```ts
import { ParsedUrlQuery } from 'querystring';

const p: ParsedUrlQuery = {
    // This shouldn't be allowed
    x: undefined,
};
```

These problems occur whether or not you use `noUncheckedIndexedAccess`.

|                                                                                                                  | keep `undefined` | remove `undefined` |
| ---------------------------------------------------------------------------------------------------------------- | ---------------- | ------------------ |
| indexing returns  when `noUncheckedIndexedAccess` is disabled `undefined`                                                                                     | ✅               | ❌                 |
| indexing returns `undefined` when `noUncheckedIndexedAccess` is enabled                                          | ✅               | ✅                 |
| `undefined` is not part of the index signature when it's used outside of indexing | ❌               | ✅                 |

It's a trade off, so I guess we have to decide which set of users we want to optimise these types for.

Personally, I've had a great experience using `noUncheckedIndexedAccess`, so I would optimise for that use case. This is just my opinion, but now we have a flag for the old problem of "indexing doesn't return `undefined`", it doesn't make sense to continue to support workarounds for that problem, especially given that these workarounds just create more problems in other places.

**Comment by G-Rath** [comment]
> these workarounds just create more problems in other places.

I don't think the "problems" are as impactful as making `noUncheckedIndexedAccess` a requirement for a project, which affects the *entire* project.

Meanwhile you can very easily cast away `undefined`:

```
(Object.values(process.env) as string[]).map((value) => {
  // Expected: string;
  // Actual: string; ✅
  value;
});
```

This leaves your second example as the only "problem" I'm aware of, which while unfortunate, imo isn't that serve since it only actually impacts the `in` operator which (my experience at least) is rarely an issue.

Ultimately my point here boils down that the current layout of the types allows us to resolve these problems on a case-by-case basis using casting, whereas the new proposed layout for the types requires us to turn on a project-wide flag that has known documented edge cases that occur in very common day-to-day patterns.

**Comment by OliverJAsh** [comment]
It's a valid point. I guess we'll have to try and build a consensus about what to do here. What do other people think?

**Review by ExE-Boss — APPROVED** [review]

**Comment by ExE-Boss** [comment]
That’s why I suggested also allowing the `?` optional suffix to be valid on index signatures, so that `noUncheckedIndexedAccess` would always be enforced for that index signature.

**Comment by OliverJAsh** [comment]
@ExE-Boss Sorry, I don't follow. Can you elaborate a little bit? Sorry if I missed a comment from you somewhere.

**Comment by ExE-Boss** [comment]
That was in some issue in the [microsoft/TypeScript](https://github.com/microsoft/TypeScript) repo that I can’t find anymore.

---

The gist of it being that:
```ts
interface Dict<T> {
	[key: string]?: T;
}
```

Would cause unchecked reads of `Dict` to include `undefined` like with `noUncheckedIndexedAccess`, regardless of whether `noUncheckedIndexedAccess` is actually enabled in the project config.

**Comment by G-Rath** [comment]
> Personally, I've had a great experience using `noUncheckedIndexedAccess`, so I would optimise for that use case.
> I wouldn't like to make assumptions about how people are using the in operator. This could be a significant issue.

Likewise, I wouldn't like to make assumptions about the compatibility of peoples (and *companies*) projects are with `noUncheckedIndexedAccess`, as it could be just as a significant issue.

You can actually "enable" the desired behaviour per project using declaration merging:

```
declare global {
  namespace NodeJS {
    interface Dict<T> {
      [key: string]: T;
    }
  }
}
```

This has the exact same effect as this PR, except with the benefit of only applying to the users who opt in to the extra strictness as they do with `noUncheckedIndexedAccess`.

Because of this, I think we should remain shipping `T | undefined` as its better suited to the defaults recommended by the TypeScript team since that's what the majority of users will be using and use to from past behaviour.

**Comment by OliverJAsh** [comment]
> You can actually "enable" the desired behaviour per project using declaration merging:

If that's an option then equally, if we removed `undefined`, users could use declaration merging to add `undefined` if that's the behaviour they wanted.

However, I'm not sure if declaration _merging_ allows you to _override_ the index signature. I couldn't get this to work:

```ts
export const test = 1;

declare global {
    namespace NodeJS {
        interface Dict<T> {
            [key: string]: string;
        }
    }
}

Object.values(process.env).map((value) => {
    // Expected: string;
    // Actual: string | undefined; ❌
    value;
});
```

Can you provide a complete test case?

**Comment by G-Rath** [comment]
> If that's an option then equally, if we removed undefined, users could use declaration merging to add undefined if that's the behaviour they wanted.

That's why I said:

> Because of this, I think we should remain shipping T | undefined as its better suited to the defaults recommended by the TypeScript team since that's what the majority of users will be using and use to from past behaviour.

`noUncheckedIndexedAccess` is (currently) for users who to go [above and beyond on their strictness](https://github.com/tsconfig/bases/pull/40#issuecomment-764475433), which is why I don't think we should be treating it as the default behaviour - users who understand `noUncheckedIndexedAccess` will also be familiar 

> However, I'm not sure if declaration merging allows you to override the index signature. I couldn't get this to work:

Damn, you're right (no offense 😅) - I'm guessing my IDE was still waking up, as it's now reporting errors no matter which way I slice it 😞

**Comment by OliverJAsh** [comment]
> Because of this, I think we should remain shipping T | undefined as its better suited to the defaults recommended by the TypeScript team

The current workaround of adding `undefined` to the index signature adds strictness. You're arguing that we should keep this workaround for users who don't want to enable this flag _because the flag is too strict_? That seems like a contradiction.

If someone really cares about strictness when indexing, they would enable the flag. If it's a large migration there are many ways they can migrate incrementally. If they don't care about strictness, well, they have nothing to worry about—they won't need this workaround nor the flag.

**Comment by G-Rath** [comment]
> The current workaround of adding undefined to the index signature adds strictness

It adds a different kind of strictness: requiring *either* explicit checks *or* explicit definition of properties in the types.

We're talking about different levels of strictness here - my reasoning is that properties like `process.env` and `headers` on http requests are "input boundary properties", meaning they come from external sources that don't have TypeScript ensuring they have required properties.

Hence, its reasonable for them to default to being "x *or* not defined", given that you can define them by using declaration merging:

```
declare module 'aws-lambda' {
  export interface APIGatewayProxyEventHeaders {
    // these will always be there, because I've got external validations in API Gateway for these headers
    'X-GitHub-Event': GithubEventName;
    'X-Hub-Signature-256': string;
  }
}

declare global {
  export namespace NodeJS {
    export interface ProcessEnv {
      // this will always be there as I've got external validations for it in my serverless configuration
      GITHUB_WEBHOOK_SECRET: string;
    }
  }
}

type ApiEvent = Pick<APIGatewayProxyEvent, 'body' | 'headers'>;

export const getEvent = (event: ApiEvent): GithubEvent => {
  const rawBody = event.body ?? '{}';
  const signature = event.headers['X-Hub-Signature-256'];
  const { GITHUB_WEBHOOK_SECRET } = process.env;

  if (verify(GITHUB_WEBHOOK_SECRET, rawBody, signature)) {
    // ensure we're referencing a header that is defined, outside of the cast
    const name = event.headers['X-GitHub-Event'];
    const payload = JSON.parse(rawBody) as GithubEvent['payload'];

    return { name, payload } as GithubEvent;
  }

  throw new Error('event did not come from github');
};
```

> (In fact, the current workaround would probably be undesirable for them, since it forces strictness.)

As I've shown above, there are multiple safe & reasonable methods to work with this, including one that doesn't require you to do any casting. This is not the case for `noUncheckedIndexedAccess`.

> If someone really cares about strictness when indexing, they would enable the flag so they can benefit from strictness not just when indexing Node types but all types e.g. arrays

But the flag has known failings with common patterns that means it's reasonable to both care about strictness and not have the flag enabled. This is also a big part of my underlying point: the flag affects *all* index checks, not just objects - so requiring users to turn it on to regain a level of strictness which we've already achieved with minimal issue isn't a good move imo.

If we had multiple flags that let us have strictest indexing on objects but not arrays, I'd be a lot more open to this change of the default, but we don't - it's all or nothing, which greatly increases the impact of this change.

> If it's a large migration there are many ways they can migrate incrementally

The only way to incrementally migrate that I know of is would be heavy use of unstrict and unsafe casts and `@ts-expect-error` comments.

-----

Just to be clear, I'm not against the flag itself as I think its a step in the right direction - I just think we should cater towards TypeScripts recommendations on what flags should be enabled by default, as I think they've got a good understanding of the potential impacts enabling a flag could have on wide-range of codebases that use TypeScript (I'd say the best, but suspect some people would fight me over that).

I also think that having the behaviour of this interface change in this manner when you've on TS >= 4.1 is dangerous, as it's a very subtle change to a very common interface which is used by a number of properties; you're assuming everyone who upgrades will jump on enabling the flag which I don't think is the case (in particular because the "what's new" notes say "might be noisy for a lot of code, so it is not automatically enabled by the --strict flag.", making it a prime target for "lets not enable this on our huge codebase").

Just because *you've* had a nice experience with the flag, doesn't mean others have or will, and nor does it mean projects can afford to take the hit that comes with enabling this flag. We'd all love to be as strict as possible in every project we do, but we have to balance that with things like time, budgets, morale, etc. The current structure of the types gives us a middle-ground that lets us be strict about externally enforced/controlled data without requiring us to make changes across the board (the specifics of which are unknown, since the flag is very new and we've not all had a chance to play with it). 

I'm not saying this change should never be made to these types either - just that they shouldn't be made now. My hope is that we'll see improvements to both this flag & TS in general that'll make it smarter to a point where it can be recommended as a default, but for now this is what we've got.

**Comment by OliverJAsh** [comment]
I don't have the time to properly respond at the moment but I just want to say before it's too late: it would be good to get some more input from other maintainers so we can reach some consensus here on the best way forward. The PR is approved by an owner but we should get more input before deciding whether to merge or not. There are many trade offs and our decision should represent the majority of users rather than the needs of a few.

**Comment by Flarna** [comment]
I tried `noUncheckedIndexedAccess` in one of my existing projects and it results in a lot errors issued by typescript.
I didn't look into all failures but quite some of them seemed to be wrong requiring to add a lot of casts or `!` or unneeded `if` checks where I have to ask myself what to write into the `else` as it can't happen anyway.

So the effort to switch this on in existing projects seems to be quite high. For now I would prefer to have definitions which work fine even with `noUncheckedIndexedAccess` set to `false`.

**Comment by OliverJAsh** [comment]
@Flarna Thanks for sharing. Just to clarify in case it wasn't clear, the types would still work without `noUncheckedIndexedAccess`, you would just lose a bit of strictness i.e. indexing would behave like it does outside of Node types.

**Comment by Flarna** [comment]
> you would just lose a bit of strictness

The idea of `noUncheckedIndexedAccess` was to have more strictness. But looking in this PR it seems the result is the opposite as I assume most users have it not enabled (as it is disabled on default after calling `tsc --init` and they now loose a bit of strictness.

**Comment by OliverJAsh** [comment]
That's correct—only `noUncheckedIndexedAccess` users would have the stricter behaviour, which would bring the Node types inline with the how index signatures are usually defined in other places (without `undefined`), e.g. `PropertyDescriptorMap` in `lib.es5.d.ts`.

I just realised that some of the index signatures defined in TypeScript's `lib`s also use this workaround, so I posted an issue in the TypeScript repo to ask whether they plan to remove these workarounds. I hope the TypeScript's team answer in that issue will provide some guidance on the best way forward. https://github.com/microsoft/TypeScript/issues/42810

**Comment by Flarna** [comment]
In my opinion the needed feature here is some better support in typescript to allow definition of interfaces like now if `noUncheckedIndexedAccess` is not set but at the same time all to get the features if it is set.

**Comment by OliverJAsh** [comment]
Closing until we get some feedback in https://github.com/microsoft/TypeScript/issues/42810.

**Pull request closed without merging**
